[addons] Use changelog.txt as an alternative to news XML tag

AlwinEsch opened the pull request:
## Description

In the event that the changes log (news) in addon.xml is empty, check whether it is an installed addon and read a changelog.txt as a replacement, if available.

Many addons use this "changelog.txt", but less use the `<news>` field in addon.xml, and nobody actually uses the translation path in it (once unknown and difficult to keep correct in the case of continuous changes).
Hence, this brings a fall back to reading the text file as a replacement.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context

This is second pull request try, by them before https://github.com/xbmc/xbmc/pull/20209 was Jenkins strange.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

Here about this file https://github.com/xbmc/audiodecoder.sacd/blob/Matrix/audiodecoder.sacd/changelog.txt given to it:
![Bildschirmfoto von 2021-09-30 14-39-01](https://user-images.githubusercontent.com/6879739/135457690-f9491d8e-c958-4c15-b9a3-72214083c21b.png)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Sorien commented:
what about extending news with an attribute like source `<news source="changelog.txt"/>`?

enen92 commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -360,8 +367,26 @@ bool CAddonInfoBuilder::ParseXML(const AddonInfoPtr& addon, const TiXmlElement*
+          CFile file;
+          auto_buffer buf;
+          if (file.LoadFile(changelog, buf) > 0)
+            addon->m_changelog["en_GB"].assign(buf.get(), buf.size());
```
can we get the `en_GB` string (as the default language) from somewhere? Or define it beforehand on an anonymous namespace?

AlwinEsch replied:
My English is on strike, what do you mean exactly?
Allow other languages there and make them available by means of a function?

E.g.:
```
addon->m_changelog[GetLanguage()].assign(buf.get(), buf.size());
```

Wouldn't make sense then, because no translations are possible there on text file.

enen92 replied:
I mean simply to define the en_GB string somewhere and reuse it as KODI_DEFAULT_LANGUAGE or something similar. Someone reading the code (and not familiar with our defaults) probably can't figure out why we are only setting the British English value. Another option is probably to get it from some central place (so that if we change defaults it's reflected everywhere)

AlwinEsch replied:
Have added a new commit about where change all related places to it.

enen92 commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -116,6 +117,12 @@ bool CAddonInfoBuilder::ParseXML(const AddonInfoPtr& addon, const TiXmlElement*
+   * The function variable "repo" is only used when data stored on the internet
+   * are read out. A boolean value is then set here for easier identification.
+   */
+  const bool isRepoXMLContent = !repo.artdir.empty() && !repo.datadir.empty();
```
isn't the `!repo.datadir.empty();` enough here? Or should we really check if the art dir is empty? art seems to have no relation with this

AlwinEsch replied:
Should be enough, there was just a little paranoia on my mind.

AlwinEsch commented:
> what about extending news with an attribute like source `<news source="changelog.txt"/>`?

There comes the human factor about, if it by e.g. this, then a developer can overlook it and forget about.

I also considered something like that, but thought to bring it in when my planned Markdown support comes and with support for.
E.g. as this:
- `<news  lang="de_DE" source="resources/language/resource.language.de_de/changelog.md"/>`
- `<description  lang="en_GB" source="resources/language/resource.language.en_gb/description.md"/>`
- `<description  lang="auto" source="resources/language/resource.language.${LANG}/description.md"/>`

But with that a way has to be found to have it here correctly https://kodi.tv/addons/matrix/audiodecoder.hvl and a way how something like that can be translated at weblate.

The only thing it would make sense for would be if a developer wanted to save it with a different name or in different locations, but the mirror problem of addons also comes back.

To have as example on here `http://mirrors.kodi.tv/addons/matrix/audiodecoder.asap+android-aarch64/changelog.txt` the last updated one, can it be found easily, but if his place/name different becomes it more hard.

AlwinEsch commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -324,7 +331,8 @@ bool CAddonInfoBuilder::ParseXML(const AddonInfoPtr& addon, const TiXmlElement*
       if (element && element->GetText() != nullptr)
       {
         addon->m_lifecycleState = AddonLifecycleState::BROKEN;
-        addon->m_lifecycleStateDescription.emplace("en_gb", element->GetText());
```
Here was also a typo about upper/lowercase :sweat_smile:.

enen92 reviewed (approved):
Thanks

enen92 commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -116,6 +117,12 @@ bool CAddonInfoBuilder::ParseXML(const AddonInfoPtr& addon, const TiXmlElement*
   }
 
+  /*
+   * The function variable "repo" is only used when data stored on the internet
```
*when reading data stored on the internet.

enen92 commented on xbmc/addons/addoninfo/AddonInfoBuilder.cpp:
```diff
@@ -360,8 +368,26 @@ bool CAddonInfoBuilder::ParseXML(const AddonInfoPtr& addon, const TiXmlElement*
-      /* Parse addon.xml "<news lang="..">...</news>" */
+      /* Parse addon.xml "<news lang="..">...</news>"
+       *
+       * In the event that the changes log (news) in addon.xml is empty, check
```
*that the changelog

enen92 commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -27,6 +27,12 @@ class CAddonInfo;
 typedef std::vector<AddonInfoPtr> AddonInfos;
 
+/*!
+ * Defines the default language to use as fallback, if wanted not available.
```
*defines the default language code used as fallback in case the requested language is not available. Used, for instance, to handle content from addon.xml

enen92 reviewed (commented):
Just a few language suggestions. Thanks for providing this feature. We have quite a strong anti-news movement :)

AlwinEsch committed:
[addons] Use changelog.txt as an alternative to news XML tag

In the event that the changes log (news) in addon.xml is empty, check whether it is an installed addon and read a changelog.txt as a replacement, if available.

Many addons use this "changelog.txt", but less use the `<news>` field in addon.xml, and nobody actually uses the translation path in it (once unknown and difficult to keep correct in the case of continuous changes).
Hence, this brings a fall back to reading the text file as a replacement.

AlwinEsch committed:
[addons] use KODI_ADDON_DEFAULT_LANGUAGE_CODE define to get "en_GB"

This implements the use of the "en_GB" for use via definition (constexpr). Once to keep all passages correct and once to make it easier to understand why "en_GB".

AlwinEsch committed:
[addons] use new value isRepoXMLContent on CAddonInfoBuilder::ParseXML

The background to this is the need to distinguish between the locally installed add-on and the repo-stored add-on, so that the currently used type can be better understood.
Previously this was only possible with `repo.artdir.empty()` which made it difficult to see the background.

AlwinEsch commented:
> @AlwinEsch I renamed the PR a bit :)

Thanks about!

@enen92 have added your suggestions, thanks to look about (yeah, my english need always a review :sweat_smile:)!
